Add CLI command for extracting root xprvs

Anviking opened the pull request:
# Issue Number

#1316 


# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have added `cardano-wallet-jormungandr key root --type random <mnemonic words>`
- [x] I added unit tests for help-text and actual usage.
- [x] I added a *pending* test making sure keys are compatible with jcli (which would fail)

# Comments

```bash
$ cardano-wallet-jormungandr key --help
Usage: cardano-wallet-jormungandr key COMMAND
  Derive keys from mnemonics.

Available options:
  -h,--help                Show this help text

Available commands:
  root                     Extract root xprv as hex (64 bytes private key + 32
                           bytes chain code)
$ cardano-wallet-jormungandr key root --help
Usage: cardano-wallet-jormungandr key root --type KEYTYPE MNEMONIC_WORDS...
  Extract root xprv as hex (64 bytes private key + 32 bytes chain code)

Available options:
  -h,--help                Show this help text
  --type KEYTYPE           Any of the following:
                             random (Daedalus, 12 words)
                             icarus (15 words)
                             trezor (12, 15, 18, 21, or 24 words)
                             ledger (12, 15, 18, 21, or 24 words)
$ cardano-wallet-jormungandr key root --type random flock advance execute country leader exotic mix twenty six margin orient meat
68a0f29e6bd5d8af7ffd00a55006afa8af6fbdbded07984ddf7fb1c31c66f7460685e5d1016553fccc9724f5ee95dd8d66facd2ac1bb2f6fcd7fa5e53c97a57f50c592fcd18b67bf3393a16184d009fb25450b2de8079f870222874e804584a8
```

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -298,6 +319,144 @@ runCli = join . customExecParser preferences
+         formatEnglishEnumerationRev (ult:penult:revBeginning)
+            = intercalate ", " (reverse revBeginning)
+                ++ ", "
+                ++ penult
```
This probably doesn't work correctly for two elements. We only use and test with 1 and >2 elements.

Anviking replied:
Fixed in 3d54525

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -298,6 +319,147 @@ runCli = join . customExecParser preferences
+            icarusStyleKeyToSeed
+    Ledger ->
+        mkScheme (Proxy @'Ledger)
+            icarusStyleKeyToSeed
```
Trezor and Ledger have different seed generation styles. See: https://input-output-hk.github.io/cardano-wallet/haddock/edge/cardano-wallet-core-2020.1.27/Cardano-Wallet-Primitive-AddressDerivation-Icarus.html#v:generateKeyFromHardwareLedger

Anviking replied:
Done in 09242e8b7a6d9fd367796d2c82853e49bd37de48, although somewhat awkwardly (see https://github.com/input-output-hk/cardano-wallet/pull/1321#discussion_r377580950)

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -298,6 +319,147 @@ runCli = join . customExecParser preferences
+byronCliKeyScheme :: ByronWalletStyle -> CLIKeyScheme
+byronCliKeyScheme = \case
+    Random -> mkScheme (Proxy @'Random)
+            (Byron.getKey . flip Byron.generateKeyFromSeed pass)
```
For consistency and readability with below, I'd be better to have that as a named function as well :+1:

Anviking replied:
done.

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -298,6 +319,147 @@ runCli = join . customExecParser preferences
+
+instance (KnownNat a, ManyNatVal as) => ManyNatVal (a ': as) where
+  manyNatVal _ =
+    natVal (Proxy :: Proxy a) : manyNatVal (Proxy :: Proxy as)
```
This belongs to `Cardano.Wallet.Api.Types` next to `AllowedMnemonics`, don't you think?

Anviking replied:
6c5ec2116d98396351514ca03250a6903ae73c3e

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -298,6 +319,147 @@ runCli = join . customExecParser preferences
+
+    typeOptions = string <$>
+        [ "Any of the following:"
+        , "  random (Daedalus, " ++ allowedWords Random ++ ")"
```
We should not mention Daedalus here because this wallet scheme is not specific to Daedalus. Exchanges are using it as well. Stick with `random` :+1:

Anviking replied:
A fleeting concern I had was that one _might_ think that that `random` meant "choose one of the other three schemes at random" 🙈

Anviking replied:
10c1378dbc9f858455fde1d5bb52c5e0fa69c70e

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -242,6 +247,9 @@ data ByronWalletStyle
     | Ledger
+    deriving (Show, Generic, Eq, Bounded, Enum)
+    deriving (ToText, FromText)
+            via (BoundedEnum 'KebabLowerCase ByronWalletStyle)
```
This seems like an odd way of writing: 

```hs
instance FromText ByronWalletStyle where
    fromText = fromTextToBoundedEnum SnakeLowerCase

instance ToText ByronWalletStyle where
    toText = toTextFromBoundedEnum SnakeLowerCase
```

In order to save 2 lines, we have to introduce `DerivingVia` + a whole new type `BoundedEnum`. Please don't.
To be clear, I am not advocating against `DerivingVia` but this is not the right place for introducing this because:

a) We really aren't saving much here... The instances are already straightforward.
b) Rest of the code using the casing functions don't use DerivingVia. So, if something like this is to be introduced, then it needs to be done consistently across all call-site where `SnakeLowerCase` are used. Else, we now have 2 different ways of maintaining parts of code that are doing the exact same thing. 
c) This PR isn't the right place for going through b).

Anviking replied:
> We really aren't saving much here... The instances are already straightforward.

Personally I find:
- It's easier to see that the `instance` declarations belong to `ByronWalletStyle` as a `deriving` statement directly under the type declaration. Multiple instances following the type declaration make it more difficult to see the structure in the module.
- `DerivingVia` guarantees that `ToText` and `FromText` are consistently derived using the same mechanism. I.e both via `BoundedEnum` and both using `KebabLowerCase`.

> then it needs to be done consistently across all call-site where

I thought some diversity / incremental-ity could be fine, but sure.

>  but this is not the right place for introducing this because:

Fair point.

KtorZ replied:
I agree with both of your first point and again, I am not against using DerivingVia. But, if we use `A`, we do it consistently. Merging this as it is would introduce `B` and keep `A` in all other places. This makes the code less maintainable. The key for maintainability is consistency. If patterns are not obvious, then we will inevitably deviate from existing patterns and soon, there will be a `C` way of doing it.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -707,3 +711,20 @@ hex = convertToBase Base16
+
+-- | Convert a key as hex-encoded text
+keyToHexText :: XPrv -> Text
+keyToHexText =
```
This is a very misleading name for a function that actually modifies the internals of the `XPrv`! 
Plus, this function is trying to do 2 things at once:

- Stripping the public part
- Converting to hex

We already have a function that convert ByteString to hexadecimal, make use of it! It's declared 10 lines above!
Now, what about writing a function `stripPubKey :: ByteString -> ByteString` and at call-site, compose them all!

```hs
let bytes = hex . stripPubKey . unXPrv $ xprv 
```

Much clearer and honest about what's being done!

Anviking replied:
👍 Went with `unXPrvStripPub` which you mentioned as an alternative in DM

>  compose them all!

`stripPubKey` is a bit too composable for my liking in this case. `stripPubKey . stripPubKey` makes no sense.

Here I don't see any benefit of composability. In fact, it would also be nice to be keep the jcli-compatibility test as similar to the actual usage in the CLI as possible.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivationSpec.hs:
```diff
@@ -211,6 +215,14 @@ spec = do
         it "XPub IcarusKey"
             (property $ prop_roundtripXPub @IcarusKey)
 
+    xdescribe "keyToHexText (not actually compatible with jcli)" $ do
```
:thinking: ?

Anviking replied:
Stripping the public key in `keyToHexText` was done to match our format with that of jcli. It didn't get us all the way  https://input-output-rnd.slack.com/archives/CCPKJ93Q9/p1580396136029600

You probably think we shouldn't include it as a pending test?

KtorZ replied:
But certainly it does work for keys that are rightfully generated (i.e. with the ed25519 bits tweak)? The Arbitrary instance you use in the test have encryption passphrases, so in most case, you're not trying to import a key, but its encrypted version (which likely is an invalid key). 

This test works just fine if you use empty passphrases / unencrypted keys. 

```hs
keyToHexText (not actually compatible with jcli)
    is compatible with jcli (Shelley)
      +++ OK, passed 100 tests.
```

Anviking replied:
🤨 Hmm…

Anviking replied:
Fixed.

jonathanknowles commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -298,6 +322,133 @@ runCli = join . customExecParser preferences
+parseSeedWords = some (argument str (metavar "MNEMONIC_WORDS..."))
+
+-- | Lay thine eyes upon the type parameters and see that there are none.
+data CLIKeyScheme = CLIKeyScheme
```
Nit-pick suggestion: use PascalCase here, so:

```suggestion
data CliKeyScheme = CliKeyScheme
```

We then achieve consistency with other functions, such as:
```hs
 byronCliKeyScheme :: ByronWalletStyle -> CliKeyScheme
```

jonathanknowles commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -298,6 +322,133 @@ runCli = join . customExecParser preferences
+    -- argument.
+    mkScheme _ = CLIKeyScheme
+            (left getFromMnemonicError . fromMnemonic @(AllowedMnemonics s) @"seed")
+            (map fromIntegral (manyNatVal $ Proxy @(AllowedMnemonics s)))
```
Somehow this indent (and others like it) ended up being 8 spaces instead of 4 spaces.

Perhaps reduce this back to 4 spaces?

jonathanknowles commented on lib/cli/test/unit/Cardano/CLISpec.hs:
```diff
@@ -473,6 +533,44 @@ spec = do
+        (["key", "root", "--type", "random"]) `expectStdErr`
+            (`shouldStartWith` "Missing: MNEMONIC_WORDS...")
+
+        let shrug = "¯\\_(ツ)_/¯"
```
:laughing:

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivationSpec.hs:
```diff
@@ -279,6 +292,23 @@ prop_passphraseHashMalformed
+  where
+    jcliKeyFromHex = readProcessWithExitCode
+            "jcli"
+            ["key", "from-bytes", "--type", "ed25519bip32"]
```
Another mysterious 8 character indent! :wink:

Anviking replied:
Not sure, I might be affected by multiline let-bindings inside do-notation where 4-characters results in parse errors 🤷‍♂

Anviking commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -388,43 +388,83 @@ parseSeedWords = some (argument str (metavar "MNEMONIC_WORDS..."))
+    byronKeyFromSeed = (Byron.getKey . flip Byron.generateKeyFromSeed pass)
+
+    hardwareKeyFromMnemonic :: [Text] -> Either String XPrv
+    hardwareKeyFromMnemonic mw = Icarus.getKey <$> case length mw of
```
This is now very ugly 🤔 
- It is duplicated with the API.
- It does **not** respect the `AllowedMnemonics`.

Anviking replied:
We'd need something like
https://github.com/input-output-hk/cardano-wallet/blob/09242e8b7a6d9fd367796d2c82853e49bd37de48/lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs#L392-L401

Maybe:
- we can return some `forall mw. (Mnemonic mw)` instead of `Passphrase purpose` (in fromMnemonic)
- we can adjust some existing pieces (like `generateKeyFromHardwareLedger`) to work better together somehow
- we can remove the `mw` type parameter on `Mnemonic mw`. I've grown more and more questioning of it at least.

Anviking replied:
Resolved with #1343

Anviking commented on lib/cli/test/unit/Cardano/CLISpec.hs:
```diff
@@ -473,6 +533,44 @@ spec = do
+            \9eb7ad13f798d06ce550a9f6c48dd2151db4593e67dbd2821d75378c7350f1366b\
+            \85e0be9cdec2213af2084d462cc11e85c215e0f003acbeb996567e371502\n"
+        (["key", "root", "--type", "random"] ++ mw12) `shouldStdOut`
+            "f0ca39a0f8e259704de6fe79b61926411515b97ec99c265a55aefb5c2094e25a80\
```
I've never confirmed these goldens with anything. It might be neat to guarantee that deriving address keys manually via the CLI match the unused addresses in the API for new wallets.

Would have to be done when we have `key child --path` though.

KtorZ replied:
Where do the values come from then?

Anviking replied:
From the currently implemented CLI itself. 

I.e. they are not guaranteed to be correct, but they won't accidentally change.

KtorZ replied:
It seems very wrong that the `trezor` and `ledger` keys are equal.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -246,6 +260,20 @@ type instance AllowedMnemonics 'Ledger    = '[12,15,18,21,24]
+
+instance (KnownNat a, ManyNatVal as) => ManyNatVal (a ': as) where
+  manyNatVal _ =
+    natVal (Proxy :: Proxy a) : manyNatVal (Proxy :: Proxy as)
```
I just remembered that this already exists as `NatVals` in module Cardano.Wallet.Primitive.AddressDerivation

Anviking commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivationSpec.hs:
```diff
@@ -214,6 +219,14 @@ spec = do
         it "XPub IcarusKey"
             (property $ prop_roundtripXPub @IcarusKey)
 
+    describe "keyToHexText" $ do
```
Function was renamed to `unXPrvStripPub`. Should update.

Anviking commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -296,6 +323,156 @@ runCli = join . customExecParser preferences
+parseSeedWords = some (argument str (metavar "MNEMONIC_WORDS..."))
+
+-- | Lay thine eyes upon the type parameters and see that there are none.
+data CliKeyScheme = CliKeyScheme
```
It might be a good idea to move most of this, as well as `unXPrvStripPub` to a separate module in `cardano-wallet-cli`.

Anviking replied:
_also, that comment could be a bit more helpful_

Anviking replied:
- Leaving in the current module for now.
- Comment for `CliKeyScheme` has been updated.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -722,3 +728,28 @@ hex = convertToBase Base16
+--
+-- TODO: I'm not quite sure how this works, despite rountrip test passing.
+-- Sanity check.
+xPrvFromStrippedPubXPrv :: ByteString -> Either String XPrv
```
This confused me:
```haskell
> bytes = "(\134\242|I\141L\EM\NUL\128\173\252q\191\172\167>f \218\222\167.\136\DC4\216\191\253r8cD8&I\STX;\185&\177\172E\241\185\241\157\226\r\163+\EM\GS\232-\188\250[E^N\129J\158\STX\134\188!\241\DLEzt\222\199\247U\143\ETB\128,\226Q\"\230\234\"\191\177\250\230\167\n\214X\244z\\"
> f = unXPrvStripPub . (either error Prelude.id . xPrvFromStrippedPubXPrv)
> f (BS.take 2 bytes) == bytes
True
-- However:
> BS.take 2 bytes
"(\134"
> f "(\134" == bytes
False
```
At first it seemed like I was magically able to re-create the XPrv from 2 bytes. But then it seemed like there is some behind the scenes magic related to `BS.take`.

Anviking replied:
This is interesting also 😅 
```haskell
>>> unXPrv <$> xPrvFromStrippedPubXPrv (BS.take 0 bytes)
fish: 'stack ghci cardano-wallet-core' terminated by signal SIGSEGV (Address boundary error)
```

Anviking replied:
https://github.com/input-output-hk/cardano-crypto/issues/67

KtorZ replied:
@Anviking lazyness maybe? Nothing tells you that you have to reduce your bytestring to WHNF for comparing them, perhaps the underlying implementation looks at length first, which is known to be different, so the Bytestring has to be different :man_shrugging:

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -722,3 +728,28 @@ hex = convertToBase Base16
+    $ CC.encryptedCreateDirectWithTweak x pass
+  where
+    pass :: ByteString
+    pass = ""
```
It'd be better to have this as a function argument and pass an empty passphrase at the call-site! Avoid hard-coding things in functions.

Anviking replied:
In that case we'd have:
```haskell
xPrvFromStrippedPubXPrv :: Passphrase "encryption" -> ByteString -> Either ErrXPrvFromStrippedPubXPrv XPrv
```
taking a password, but
```haskell
unXPrvStripPub :: XPrv -> ByteString
```
not taking a password.

But I guess this is fine as the password would already be "embedded" in the `XPrv`.

That said I'm not quite sure these functions should work regarding passwords. The following test fails:

```haskell
prop_strippedPubRoundtrip
    :: Unencrypted (ByronKey 'RootK XPrv)
    -> Passphrase "encryption"
    -> Property
prop_strippedPubRoundtrip (Unencrypted k) pwd = do
    let f = (xPrvFromStrippedPubXPrv pwd) . unXPrvStripPub . (xPrvChangePass empty pwd) . getRawKey
    let raw = unXPrv . getRawKey
    either (error . show) unXPrv (f k) === raw k
  where
    empty = Passphrase ""
```
because the public keys are different (prv <> pub <> cc) 🤔 
```
"e70d773a098fc7f290b335851a2d7bc5f5a887e97f5b9e7691a414ad04ffbea3d4df6900b462e1e0f9ffeda60d42f25e35555b9059f2a9a4c1c7c019d0efbe677ddaa81e31d0b2c1a939221cf94268e2209d09e191f1fe10ec5a4662cdb1e09caa420eff34b0baf7afccf387517ac43cc2a8059204571bd94808d0e5cb917c14"
/=
"e00d773a098fc7f290b335851a2d7bc5f5a887e97f5b9e7691a414ad04ffbe43d4df6900b462e1e0f9ffeda60d42f25e35555b9059f2a9a4c1c7c019d0efbe6773687a4adefbd180e9aa41e0f4a06638aeee9c6bb76a03374b05f8f574462263aa420eff34b0baf7afccf387517ac43cc2a8059204571bd94808d0e5cb917c14"
```

KtorZ replied:
unXPrv would give you the **encrypted** key, so you would need to decrypt it first. It's not that the passphrase is embedded, it is that the key is encrypted. 
In principles calling `unXPrvStripPub` on a key that is encrypted should fail. Also, what exactly is the use for "xPrvFromStrippedPubXPrv" ?

Anviking replied:
> Also, what exactly is the use for "xPrvFromStrippedPubXPrv" ?

It will be needed to parse 96-byte key as a CLI argument. I.e `key child` in the future.

For now, it is only used for roundtrips.

We could move it to `Cardano.CLI`. I contemplated doing that.

KtorZ replied:
> It will be needed to parse 96-byte key as a CLI argument. I.e key child in the future.

:+1: Ok, then a passphrase makes little sense here, but using `XPrv` feels also misleading because we do actually treat them and manipulate them as unencrypted key. A newtype could capture that nicely... or at the very least a comment / note explaining why the passphrase is forced to be empty. (Still, I'd add an invariant or runtime check that, when trying to xPrvStripXPub, the given key is not 
encrypted and fail loudly if it is).

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivationSpec.hs:
```diff
@@ -282,6 +299,29 @@ prop_passphraseHashMalformed
+    :: WalletKey k
+    => Unencrypted (k 'RootK XPrv)
+    -> Property
+prop_keyToHexTextJcliCompatible (Unencrypted k) = monadicIO $ do
```
This is calling `jcli` directly, so it is so-to-speak an integration test. We expect unit test to not require any 3rd party interaction so that they can be run in isolation if needed. Please, move the test :pray:

Anviking replied:
Something like `Test.Integration.Jormungandr.Scenario.CLI.Key`… even though this test is not _completely_ related to the `key` command? 🤔

Anviking replied:
We need to duplicate the `Arbitrary (Unencrypted key)`  instances to `cardano-wallet-jormungandr:integration`. But we still need to keep them here (`cardano-wallet-core:unit`) for the `unXPrvStripPub` roundtrips.

And we need to duplicate `getMnemonic` yet again because that weird type mismatch stops us from moving it to `test-utils`.

😶 🔫

KtorZ replied:
"that weird type mismatch" ?

Anviking replied:
https://input-output-rnd.slack.com/archives/GBT05825V/p1581523877011800

```
/Users/anviking/IOHK/cardano-wallet/lib/core/test/unit/Cardano/Wallet/DB/Arbitrary.hs:477:47: error:
    • Couldn't match type ‘cardano-wallet-core-2020.1.27:Cardano.Wallet.Primitive.Mnemonic.Mnemonic
                             12’
                     with ‘Cardano.Wallet.Primitive.Mnemonic.Mnemonic mw0’
      NB: ‘Cardano.Wallet.Primitive.Mnemonic.Mnemonic’
            is defined in ‘Cardano.Wallet.Primitive.Mnemonic’
                in package ‘cardano-wallet-core-2020.1.27’
          ‘cardano-wallet-core-2020.1.27:Cardano.Wallet.Primitive.Mnemonic.Mnemonic’
            is defined in ‘Cardano.Wallet.Primitive.Mnemonic’
                in package ‘cardano-wallet-core-2020.1.27’
      Expected type: Gen (Cardano.Wallet.Primitive.Mnemonic.Mnemonic mw0)
        Actual type: Gen
                       (cardano-wallet-core-2020.1.27:Cardano.Wallet.Primitive.Mnemonic.Mnemonic
                          12)
    • In the second argument of ‘(<$>)’, namely ‘genMnemonic @12’
      In the expression: Just . SomeMnemonic <$> genMnemonic @12
      In the expression: (3, Just . SomeMnemonic <$> genMnemonic @12)
    |
477 |                 , (3, Just . SomeMnemonic <$> genMnemonic @12)
    |                                               ^^^^^^^^^^^^^^^

```
when defining `genMnemonic` in the `test-utils` package.

Anviking replied:
It's moved at least. Duplication of generators is a separate issue we can live with for the moment.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -230,6 +236,13 @@ data ByronWalletStyle
+    fromText = fromTextToBoundedEnum KebabLowerCase
+
+instance ToText ByronWalletStyle where
+    toText = toTextFromBoundedEnum KebabLowerCase
```
Side-note: we use `SnakeLowerCase` everywhere else. It has no influence here because all words are single words but it'd be better to be consistent with casing on other parts..

Anviking replied:
It seemed more appropriate for the CLI… which I'll admit might be iffy when the instance itself lives with the Api-code…

Changing to `SnakeLowerCase` since it doesn't matter.

KtorZ commented on lib/cli/test/unit/Cardano/CLISpec.hs:
```diff
@@ -473,10 +558,132 @@ spec = do
+genMnemonicOfSize 21 = mnemonicToText <$> genMnemonic @21
+genMnemonicOfSize 24 = mnemonicToText <$> genMnemonic @24
+genMnemonicOfSize n = error $ "when this test was written, "
+    ++ show n ++ " was not a valid length of a mnemonic"
```
style suggestion, use LambdaCase to avoid repeating the function name 6 times :grimacing:

KtorZ commented on lib/cli/test/unit/Cardano/CLISpec.hs:
```diff
@@ -147,6 +204,19 @@ spec = do
+            , "  -h,--help                Show this help text"
+            , ""
+            , "Available commands:"
+            , "  root                     Extract root xprv as hex (64 bytes"
```
Extract root extended private key from a mnemonic sentence* 

would be better. 

a) Avoid "cryptic" abbreviation in public-facing interface. `xprv` sounds familiar to us, but users external to the project have no idea what this is about. 
b) Don't mention "hex" here, that's simply the encoding format and is relatively irrelevant (the command-line could even support multiple encoding format). 
c) It does say from where / what the key is extracted.

KtorZ commented on lib/cli/test/unit/Cardano/CLISpec.hs:
```diff
@@ -384,9 +454,24 @@ spec = do
             ]
 
+        ["key", "root", "--help"] `shouldShowUsage`
+            [ "Usage:  key root --type KEYTYPE MNEMONIC_WORDS..."
```
I am going to be slightly pedantic here but `MNEMONIC_WORDS` should be singular. The meta-type is repeated. Also, why is there an underscore between `MNEMONIC` and `WORDS` but there's none between `KEY` and `TYPE`? The underscore reads better, but it should be used consistently across the interface.

Anviking replied:
Done. `MNEMONIC_WORD...` and `KEY_TYPE`.

(Underscores are already used elsewhere in the CLI)

KtorZ commented on lib/cli/test/unit/Cardano/CLISpec.hs:
```diff
@@ -473,10 +558,132 @@ spec = do
+genMnemonic = do
+        let n = fromIntegral (natVal $ Proxy @(EntropySize mw)) `div` 8
+        bytes <- BS.pack <$> vectorOf n arbitrary
+        let ent = either (error . show) id $ mkEntropy @(EntropySize mw) bytes
```
what about `unsafeMkEntropy` ?

Anviking replied:
Indeed 👍 

That said, I'm starting to question how valuable the `unsafe...` family of functions really are. `mkEntropy`,  `EntropySize`, etc.  come from `Cardano.Wallet.Primitive.Mnemonic`, whereas `unsafeMkEntropy` now is the odd one coming from `Cardano.Wallet.Unsafe`.

I guess they make more sense when they're able to `f . g . either (error . show) id .  h`, rather than just `either (error . show) id . f` like in this specific example.

… And I think there's _something_ attractive about having a multitude of various `unsafe...` functions all in the _same_ module. Maybe it's easier to import `unsafeFoo` from `Cardano.Wallet.Unsafe` than remember where the original `foo`-function is defined  when you're writing a test.

KtorZ replied:
Note that the `unsafe` function are only convenience. We (ought to) only use them in test code for either literal values or values that are hand-crafted and known to be correct. The real question is "why is this module part of the library and not the test-suite only". And the main reason for this is just that it's shared across multiple test modules (for what the argument is worth).

Anviking replied:
> Note that the unsafe function are only convenience.

Well, I didn't (clearly) see the convenience in writing `unsafeFoo` rather than `either (error . show) id foo`, or (the hypothetical) `unsafe foo`.

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -296,6 +331,202 @@ runCli = join . customExecParser preferences
+    . unsafeFromHex
+    . B8.pack
+    . T.unpack
+    $ t
```
What about having that in `Cardano.Wallet.Unsafe` ?

Anviking replied:
I'm inclined to keep it here because:
- I want it next to `xPrvToHexText` as a dual.
- I had thoughts about making `mapKey` take a bidirectional map that can fail* `(key1 -> m key2, key2 -> m key1)`. It would not be `unsafe` in that case. (* more on that later)

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -296,6 +331,202 @@ runCli = join . customExecParser preferences
+        in
+            CliKeyScheme
+                (allowedLengths proxy)
+                (fmap hardwareKeyFromSeed . seedFromMnemonic proxy)
```
Trezor /= Ledger. Trezor uses the same key generation method as Icarus. This is incidentally why the function is called generateKeyFromHardware**Ledger** !

Anviking replied:
Ah! I see.

Renaming `hardwareKeyFromSeed` to `ledgerKeyFromSeed` then. (And using `icarusKeyFromSeed` for Trezor.)

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -296,6 +331,202 @@ runCli = join . customExecParser preferences
+        <> progDesc "Extract root xprv as hex\n\
+                    \ (64 bytes private key + 32 bytes chain code)"
+  where
+    cmd = do
```
Why not follow the same pattern as other command-line handlers? All of them have separated the parsing of the cli argument from the execution of the command. 
This may sounds artificial, but it allows for better readability, better code-reuse (as the parsers can easily be re-used) and easier refactoring options (because the code is already decoupled). 

So: 

a) Declare a new data-type to carry the argument. (e.g. `TransactionListArgs`)
b) Declare your parsers with regards to the existing convention:
    - options are postfixed with "Option" (e.g. portOption, timeRangeOption, poolGapOption, paymentOption, randomPortOption, databaseOption, addressStateOption etc. ..) and declared under the relevant section.
    - arguments are postfixed with "Argument" (e.g. walletIdArgument, epochArgument, transactionIdArgument) 
    - Options are arguments are named after they actual name, camelCased, in the command-line to reduce discrepancy and make mapping between command-line arguments and their relevant parser easier. 

In this case, `parseKeyType` should really be called `typeOption` (although, I suggest to rename this in both the command line and here to perhaps something less ambiguous like `--wallet-style` or `--style` ?)

Anviking replied:
Thanks for the explanations. I was too unfamiliar with everything here and didn't see the structure.

> although, I suggest to rename this in both the command line and here to perhaps something less ambiguous

Yeah,  `key root --wallet-style icarus` is probably clearer than `key root --type icarus` 🤷‍♂

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -296,6 +331,202 @@ runCli = join . customExecParser preferences
+        , "  ledger (" ++ allowedWords Ledger ++ ")"
+        ]
+
+    allowedWords = (++ " words")
```
I think this is something we should have near the `AllowedWords` type family declaration.
What about:

```hs
prettyAllowedWords
    :: forall style.
        ( NatVals (AllowedMnemonics style)
        )
    => Proxy style
    -> String
prettyAllowedWords _ =
    show $ natVals $ Proxy @(AllowedMnemonics style) -- TODO: pretty formatting of english enumeration
```

Anviking replied:
Eh, maybe in a separate pr please. We should maybe have a notion of `all` wallet styles also. And might be best to have at the term-level.

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -296,6 +331,202 @@ runCli = join . customExecParser preferences
+-- | Parse a key type
+--
+-- Note that we in the future might replace the type @ByronWalletStyle@ with
+-- another type, to include Shelley keys.
```
Actually yes. I don't understand why all the functions and types are limited to "Byron". Although, "Shelley" key are simply generated following the Icarus' generation style, so they are to-so-speak supported but functions should not imply that this is only a byron thing, because it clearly isn't.

Anviking replied:
>  I don't understand why all the functions and types are limited to "Byron". 

I believe you specifically said not to support `Shelley` keys here.

With that in mind, I didn't want to unnecessarily create a new type, since `ByronWalletStyle` already existed.

The parser is now called `walletStyleOption` (not mentioning byron.) Maybe it is only a matter of renaming `byronCliKeyScheme` to say `newCliKeyScheme`?

KtorZ replied:
Meh. Yes but Shelley /= Jormungandr, although Jörmungandr uses Icarus' key style so it's already covered...  to some extend, if one knows that one can use `--icarus` to generate "Jörmungandr keys". 

>  it is only a matter of renaming byronCliKeyScheme to say newCliKeyScheme?

Yes.

Anviking replied:
Done. Now `newCliKeyScheme`.

KtorZ commented on lib/cli/test/unit/Cardano/CLISpec.hs:
```diff
@@ -473,10 +558,132 @@ spec = do
+        Right _ -> property True
+        Left e -> counterexample
+            (show (length mw) ++ " words, failed with: " ++ e)
+            (property False)
```
Good properties to have. But out of curiosity, what problems do the "CliKeyScheme" solve? I fail a bit to see why this type is necessary :thinking:  ?

Anviking replied:
The first goal I had with it was to concentrate the logic related to case switching, and demotion of type-level logic to one place.

I think it captures what we're doing neatly. For instance, it is easy to write the property `prop_allowedWordLengthsAllWork` for **all** different key derivation schemes / types.

When it comes to testing I wanted to be sure that a user performing a sequence of actions via the CLI got the expected result — that nothing ended up wrong when the result of one command was used as input of another. I thought about doing a roundtrip in IO using `caputre_` (didn't go well). Instead I opted to parameterise over `key` and `m`. This allows intricate logic to happen in `mapKey`, which can be tested easily as long we're still in `Either String`.

Depending on https://github.com/input-output-hk/cardano-wallet/issues/1353#issuecomment-587370636 it might have to be altered slightly in the future. (maybe all future functionality will be the same regardless of wallet style, not sure)

Anviking replied:
> This allows intricate logic to happen in mapKey

This would favour later having a `keyArgument :: Parser Text` instead of `keyArgument :: Parser XPrv`, which you might be at odds with.

Instinctively I'd feel being able to use, test and roundtrip a `CliKeyScheme Text (Either String)` makes it worth it.

Anviking committed:
Export `NatVals` from C.W.P.AddressDerivation

Anviking committed:
Add unXPrvStripPub (& inverse) that matches jcli

- Roundtrip properties
- Integration test verifying that it works with jcli

Why:

We need to convert between 96-byte long hex-encoded bytestrings and
XPrvs when implementing the `key root` and `key child` CLI commands.

Anviking committed:
[gardening] Leverage unsafeMkEntropy in Cardano.Wallet.Gen

Unrelated to PR.

Anviking committed:
Add cli `key root` command

- Add CLI usage tests for `key` and `key root`
- Add key-root goldens (not externally validated)
- There was a local definition of expectationFailure, using quickchecks's
counterexample, which I had to rename.
I'm not sure why the existing CLI tests uses QuickCheck when they only
run once.
- Isolate the types to one place for power & elegance
- Add functionality to check CliKeyScheme equality & more powerful tests

KtorZ commented on lib/cli/cardano-wallet-cli.cabal:
```diff
@@ -69,12 +69,14 @@ test-suite unit
     , hspec
     , optparse-applicative
     , QuickCheck
+    , silently
```
A left-over unused dependency?

Anviking replied:
Actually not 😆

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -722,3 +729,41 @@ hex = convertToBase Base16
+    stripPub xprv' = prv <> chainCode
+      where
+        (prv, rest) = BS.splitAt 64 xprv'
+        (_pub, chainCode) = BS.splitAt 32 rest
```
:+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -722,3 +729,41 @@ hex = convertToBase Base16
+xPrvFromStrippedPubXPrv :: ByteString -> Either ErrXPrvFromStrippedPubXPrv XPrv
+xPrvFromStrippedPubXPrv x =
+    let
+        res = toXPrv $ CC.encryptedCreateDirectWithTweak x pass
```
This won't work for Byron keys though. The "tweak" mention here does actually change a few bits that could have a disastrous effects on legacy keys:

```
	secret_key[0] &= 248;   /* clears the bottom 3 bits */
	secret_key[31] &= 0x1F; /* clears the 3 highest bits */
	secret_key[31] |= 64;   /* set the 2nd highest bit */
```

We should instead go for `wallet_encrypted_initialize` which simply computes the public key from an already tweaked private key. Yet, this is not exported via the FFI on Haskell. So, another option would be to make this function depends on the target version (V1 vs V2) and use the appropriate creation function.

Anviking replied:
> This won't work for Byron keys though. 

What won't work? For _unencrypted_ keys, all the roundtrips are passing. The CLISpec roundtrips include Byron keys.

I think I see how we could throw in `unXPrvStripPub` if the keys are _encrypted_, like you suggested elsewhere.

KtorZ replied:
The issue has nothing to do with encryption here. Keys are a random sequence of bytes. However, some **bits** in the first and last **byte** are tweaked in Icarus and Byron keys for some security reasons, but in a slightly different way.

Looking back into the generation methods (https://gist.github.com/KtorZ/b086754a15718affcc1b5c03c12c078a#byron), the Byron approach isn't exactly the same but should work all-the-same **for a key that has already been generated**. Soooooo. Okay. Nevermind.

Anviking replied:
> Byron approach isn't exactly the same but should work all-the-same for a key that has already been generated.

Are you saying that the difference between bit-tweaking of Byron and Icarus keys shouldn't matter in this case?

If this tweak "clears the 3 highest bits" but Byron keys only have the highest bit cleared… shouldn't that modify the key? It looks like the 2nd highest bit is set, but the 3rd highest bit should differ?

Although roundtrips for non-encrypted Byron-keys haven't failed so far. I would have expected them to fail in that case.

> for a key that has already been generated

Not sure I get this.

Anviking replied:
I tested with some old mnemonics and they seem to work (have roundtripping root keys) though.

_actually, when doing Byron child derivation I'm seeing problems with roundtrips. Haven't looked close yet._

Anviking committed:
Add (ByteArrayAccess Passphrase) instance

KtorZ reviewed (approved):
:+1: (please wait after #1358 before merging, I cherry-picked one of your commit so the two branches will conflict).
